test/validation/api/ml/ml.c

```diff
@@ -0,0 +1,501 @@
+
+	CU_ASSERT_EQUAL_FATAL(odp_pool_destroy(compl_pool), 0);
+}
+
+static void compl_alloc_max(void)
+{
+	uint64_t u64;
+	odp_event_t event;
+	odp_pool_t compl_pool;
+	odp_ml_compl_pool_param_t pool_param;
+	const int num = global.ml_capa.pool.max_num;
+	odp_ml_compl_t *compl = malloc(num * sizeof(odp_ml_compl_t));
+
+	/* Create an ML job completion pool */
+	odp_ml_compl_pool_param_init(&pool_param);
+	pool_param.num = num;
+
+	compl_pool = odp_ml_compl_pool_create(COMPL_POOL_NAME, &pool_param);
+	CU_ASSERT_NOT_EQUAL_FATAL(compl_pool, ODP_POOL_INVALID);
+
+
+	for (int i = 0; i < num; i++) {
+		compl[i] = odp_ml_compl_alloc(compl_pool);
+		CU_ASSERT_NOT_EQUAL_FATAL(compl[i], ODP_ML_COMPL_INVALID);
+
+		u64 = odp_ml_compl_to_u64(compl[i]);
+		CU_ASSERT_NOT_EQUAL(u64, odp_ml_compl_to_u64(ODP_ML_COMPL_INVALID));
+
+		event = odp_ml_compl_to_event(compl[i]);
+		CU_ASSERT_EQUAL(odp_event_type(event), ODP_EVENT_ML_COMPL);
+	}
+
+	for (int i = 0; i < num; i++)
+		odp_ml_compl_free(compl[i]);
+
+	free(compl);
+
+	CU_ASSERT_EQUAL_FATAL(odp_pool_destroy(compl_pool), 0);
+}
+
+static void compl_pool_lookup(void)
+{
+	odp_pool_t pool, pool_a, pool_b;
+	odp_ml_compl_pool_param_t pool_param;
+
+	/* Create an ML job completion pool */
+	odp_ml_compl_pool_param_init(&pool_param);
+	pool_param.num = global.num_compl;
+
+	pool_a = odp_ml_compl_pool_create(COMPL_POOL_NAME, &pool_param);
+	CU_ASSERT_NOT_EQUAL_FATAL(pool_a, ODP_POOL_INVALID);
+
+	pool = odp_pool_lookup(COMPL_POOL_NAME);
+	CU_ASSERT_EQUAL(pool, pool_a);
+
+	/* Second pool with the same name */
+	pool_b = odp_ml_compl_pool_create(COMPL_POOL_NAME, &pool_param);
+	CU_ASSERT_NOT_EQUAL_FATAL(pool_b, ODP_POOL_INVALID);
+
+	pool = odp_pool_lookup(COMPL_POOL_NAME);
+	CU_ASSERT(pool == pool_a || pool == pool_b);
+
+	CU_ASSERT_EQUAL(odp_pool_destroy(pool_a), 0);
+	CU_ASSERT_EQUAL(odp_pool_destroy(pool_b), 0);
+}
+
+static int check_event_user_area(void)
+{
+	if (global.disabled)
+		return ODP_TEST_INACTIVE;
+
+	if (((global.ml_capa.load.compl_mode_mask | global.ml_capa.run.compl_mode_mask) &
+	     ODP_ML_COMPL_MODE_EVENT) &&
+	    (global.ml_capa.pool.max_uarea_size > 0))
+		return ODP_TEST_ACTIVE;
+
+	return ODP_TEST_INACTIVE;
+}
+
+static void test_ml_compl_user_area(void)
+{
+	uint32_t i;
+	void *addr;
+	void *prev;
+	odp_pool_t pool;
+	odp_ml_compl_pool_param_t pool_param;
+	uint32_t size = global.ml_capa.pool.max_uarea_size;
+	uint32_t num = global.num_compl;
+	odp_ml_compl_t compl_evs[num];
+
+	odp_ml_compl_pool_param_init(&pool_param);
+	pool_param.num = num;
+	pool_param.uarea_size = size;
+	pool = odp_ml_compl_pool_create(NULL, &pool_param);
+	CU_ASSERT_FATAL(pool != ODP_POOL_INVALID);
+
+	prev = NULL;
+	for (i = 0; i < num; i++) {
+		compl_evs[i] = odp_ml_compl_alloc(pool);
+
+		if (compl_evs[i] == ODP_ML_COMPL_INVALID)
+			break;
+
+		addr = odp_ml_compl_user_area(compl_evs[i]);
+
+		CU_ASSERT_FATAL(addr != NULL);
+		CU_ASSERT(prev != addr);
+
+		prev = addr;
```

Comment:
Test should also memset() the entire user area to see that there's valid memory behind the returned pointer.

memset(addr, 0, size);